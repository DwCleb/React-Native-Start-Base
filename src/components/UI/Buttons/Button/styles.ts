import styled from 'styled-components/native';

import { Button } from 'react-native-paper';

import { Metrics, Colors } from '~/theme';
import { ThemeType } from '~/theme/theme';

export enum ButtonMode {
  TEXT = 'text',
  OUTLINED = 'outlined',
  CONTAINED = 'contained',
}

export interface ButtonProps {
  title: string;
  onPress(): void;
  accessibilityLabel?: string;
  icon?: string;
  color?: string;
  mode?: ButtonMode;
  loading?: boolean;
  disabled?: boolean;
  uppercase?: boolean;
  small?: boolean;
  distance?: number[];
  ratio?: number;
}

export type Props = ButtonProps & { theme: ThemeType };

function getMargin({ distance }: Props): string {
  if (!distance) {
    return `${10}px ${5}px ${10}px ${5}px`;
  } else {
    switch (distance.length || 0) {
      case 1:
        return `${distance[0]}px`;
      case 2:
        return `${distance[0]}px ${distance[1]}px`;
      case 3:
        return `${distance[0]}px ${distance[1]}px ${distance[2]}px`;
      case 4:
        return `${distance[0]}px ${distance[1]}px ${distance[2]}px ${distance[3]}px`;
      default:
        return '0px';
    }
  }
}

export const MyButton = styled(Button).attrs(
  ({
    icon,
    color = Colors.palette.primary.blue,
    mode = ButtonMode.CONTAINED,
    uppercase = false,
    loading,
    ratio = 0.85,
  }: ButtonProps) => ({
    icon,
    color,
    mode,
    uppercase,
    contentStyle: { height: 40, width: Metrics.screen.width * ratio },
    loading,
  })
)`
  margin: ${(props: Props): string => getMargin(props)};
  justify-content: center;
  align-items: center;
`;
