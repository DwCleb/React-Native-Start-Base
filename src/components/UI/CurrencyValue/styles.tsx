import Typography from '../Typography';
import styled from 'styled-components/native';
import { scale, ThemeContextType } from '~/theme';

interface WrapperProps {
  color?: string;
  opacity: string;
  valueWeight: string;
  isCurrency?: boolean;
}

type ThemedProps = ThemeContextType & WrapperProps;

function getTitleColor({ theme, color }: ThemedProps): string {
  return color ? theme.currency.title[color] : theme.monochromatic.white;
}

export const LineTroughContainer = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: center;
  position: relative;
  margin-right: ${scale(3)}px;
`;

export const PercentageContainer = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

export const Value = styled(Typography)`
  color: ${(props: ThemedProps): string => `${getTitleColor(props)}${props.opacity === 'half' ? '7F' : ''}`};
  ${({ valueWeight }: ThemedProps): string => valueWeight && `font-weight: ${valueWeight}`};
  letter-spacing: ${(props: ThemedProps): string => (props.isCurrency ? `${scale(1.5)}` : `${scale(1)}`)}px;
`;

export const LineThrough = styled.View`
  width: 100%;
  height: ${scale(2)}px;
  background-color: yellow;
  position: absolute;
  bottom: auto;
`;

function padding(prop: number[] = []): string {
  switch (prop.length || 0) {
    case 1:
      return `${scale(prop[0])}px`;
    case 2:
      return `${scale(prop[0])}px ${scale(prop[1])}px`;
    case 3:
      return `${scale(prop[0])}px ${scale(prop[1])}px ${scale(prop[2])}px`;
    case 4:
      return `${scale(prop[0])}px ${scale(prop[1])}px ${scale(prop[2])}px ${scale(prop[3])}px`;
    default:
      return '0px';
  }
}

export const Container = styled.View`
  flex-direction: ${({ row }: { row: string }): string => (row ? 'row' : 'column')};
  align-items: center;
  justify-content: flex-start;
  padding: ${({ distance }: { distance: [] }): string => padding(distance)};
`;
