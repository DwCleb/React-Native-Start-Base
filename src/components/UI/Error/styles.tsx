import styled from 'styled-components/native';
import { scale, ThemeContextType } from '~/theme';
import Typography, { TypographyType } from '~/components/UI/Typography';
import Modal from 'react-native-modal';

//Modal
export const ErrorModal = styled(Modal)`
  margin: 0px;
`;

//Parent Component
export const MainContainer = styled.SafeAreaView`
  flex: 1;
  width: 100%;
  background-color: ${({ theme }: ThemeContextType): string => theme.background};
`;

//Content's ScrollView
export const Scroll = styled.ScrollView.attrs(() => ({
  keyboardShouldPersistTaps: 'handled',
  showsVerticalScrollIndicator: false,
  contentContainerStyle: { flex: 1, paddingTop: scale(35) },
}))`
  height: auto;
`;

//Content Container
export const ContentContainer = styled.View`
  flex: 1;
  margin: ${scale(30)}px;
`;

//Page Title
export const Title = styled(Typography).attrs(() => ({
  type: TypographyType.H2,
  weight: 'regular',
  align: 'left',
}))`
  color: ${({ theme }: ThemeContextType): string => theme.text.black};
  margin-bottom: ${scale(10)}px;
  letter-spacing: ${scale(1)}px;
`;

export const SubTitle = styled(Typography).attrs(() => ({
  type: TypographyType.H5,
  weight: 'regular',
  align: 'left',
}))`
  color: ${({ theme }: ThemeContextType): string => theme.text.black};
  letter-spacing: ${scale(1)}px;
`;

//Error message
export const ErrorMessageContainer = styled.View`
  border-radius: ${scale(8)}px;
  height: ${scale(100)}px;
  padding: ${scale(20)}px ${scale(40)}px;
  justify-content: center;
  align-items: center;
  background-color: ${({ theme }: ThemeContextType): string => theme.error.card.background};
`;

export const ErrorMessage = styled(Typography).attrs(() => ({
  type: TypographyType.P2,
  weight: 'regular',
  align: 'center',
  numberOfLines: 2,
}))`
  color: ${({ theme }: ThemeContextType): string => theme.monochromatic.black};
  letter-spacing: ${scale(1)}px;
`;

//Buttons container
export const ButtonsContainer = styled.View`
  margin-bottom: ${scale(30)}px;
`;
