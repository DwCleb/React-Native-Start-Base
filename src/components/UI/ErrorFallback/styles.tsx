import styled from 'styled-components/native';
import { scale } from '~/theme';

import { Logo } from '~/assets';

export const ErrorImage = styled.Image.attrs(() => ({
  source: Logo,
  resizeMode: 'contain',
  align: 'center',
  marginTop: scale(30),
  marginBottom: scale(30),
  alignSelf: 'center',
}))`
  width: ${scale(82)}px;
`;
