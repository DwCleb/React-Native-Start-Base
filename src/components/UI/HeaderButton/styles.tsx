import styled from 'styled-components/native';

/** Theme */
import { scale, ThemeContextType } from '~/theme';

/** Assets */
import { BackIcon as BackIconAsset, CrossIcon as CrossIconAsset } from '~/assets';

/* Dependecies */
import { I18nManager } from 'react-native';

enum ButtonEnum {
  Back = 'back',
  Close = 'close',
}

interface IconProps {
  type: ButtonEnum;
}

export const ButtonContainer = styled.TouchableOpacity.attrs(() => ({
  hitSlop: { top: 30, left: 30, bottom: 30, right: 30 },
}))`
  margin-top: 0px;
  margin-left: ${(props: IconProps): string => (props.type === ButtonEnum.Back ? `${scale(20)}px` : '0px')};
  padding-right: ${(props: IconProps): string => (props.type === ButtonEnum.Back ? `${scale(30)}px` : '0px')};
`;

export const BackIcon = styled.Image.attrs((props: IconProps) => ({
  source: props.type === ButtonEnum.Back ? BackIconAsset : CrossIconAsset,
  resizeMode: 'contain',
}))`
  transform: scaleX(${I18nManager.isRTL ? -1 : 1});
  tint-color: ${(props: ThemeContextType): string => `${props.theme.navigation.back}`};
  height: ${(props: IconProps): string => (props.type === ButtonEnum.Back ? `${scale(16)}px` : `${scale(26)}px`)}
  width: ${(props: IconProps): string => (props.type === ButtonEnum.Back ? `${scale(9)}px` : `${scale(26)}px`)}
`;
