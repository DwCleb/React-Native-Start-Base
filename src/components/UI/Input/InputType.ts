export enum InputStatus {
  Valid = 'valid',
  Complete = 'complete',
  Initial = 'initial',
  Invalid = 'invalid',
}
