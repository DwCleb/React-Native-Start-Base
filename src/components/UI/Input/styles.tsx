import styled from 'styled-components/native';
import { TextInput } from 'react-native-paper';

/* Theme */
import { Metrics, scale, ThemeContextType } from '~/theme';
import { ThemeType } from '~/theme/theme';

export enum InputMode {
  OUTLINED = 'outlined',
}

export enum InputMaskType {
  CPF = 'cpf',
  CNPJ = 'cnpj',
  CREDIT_CARD = 'credit-card',
  CEL_PHONE = 'cel-phone',
  DATETIME = 'datetime',
  MONEY = 'money',
  ZIP_CODE = 'zip-code',
  ONLY_NUMBERS = 'only-numbers',
  CUSTOM = 'custom',
}

export enum PassIcon {
  OFF = 'eye-off-outline',
  ON = 'eye-outline',
}

export interface InputProps {
  distance?: number[];
  mode?: InputMode;
  paddingHorizontal?: number;
  height?: number;
  label?: string;
  helperText?: string | boolean;
  value: string;
  name: string;
  pass?: boolean;
  hasError?: boolean;
  onChange({ value, name }: { value: string; name: string }): void;
  options?: {};
  mask?: InputMaskType;
  onBlur?: () => void;
  onFocus?: () => void;
}

export type Props = InputProps & { theme: ThemeType };

function getMargin({ distance }: Props): string {
  if (!distance) {
    return `${10}px ${5}px ${10}px ${5}px`;
  } else {
    switch (distance.length || 0) {
      case 1:
        return `${distance[0]}px`;
      case 2:
        return `${distance[0]}px ${distance[1]}px`;
      case 3:
        return `${distance[0]}px ${distance[1]}px ${distance[2]}px`;
      case 4:
        return `${distance[0]}px ${distance[1]}px ${distance[2]}px ${distance[3]}px`;
      default:
        return '0px';
    }
  }
}

export const InnerInput = styled.View`
  width: ${({ paddingHorizontal }: Props): number => Metrics.screen.width - scale(paddingHorizontal || 20)}px;
  margin: ${(props: Props): string => getMargin(props)};
`;

export const InputBase = styled(TextInput).attrs(({ theme, mode, hasError }: Props) => ({
  theme: { colors: { primary: hasError ? theme.alerts.red : theme.primary.blue } },
  mode,
  autoCorrect: false,
  autoCapitalize: 'none',
  autoCompleteType: 'off',
  multiline: false,
}))`
  flex: 1;
  /* height: ${({ height }: Props): number => height || 45}px; */
  justify-content: center;
  padding-top: 0px;
  padding-bottom: 0px;
`;

export const ShowPass = styled.TouchableOpacity.attrs(() => ({
  activeOpacity: 0.7,
  hitSlop: { top: 20, bottom: 20, left: 20, right: 20 },
}))`
  left: -10px;
  align-self: flex-end;
  top: ${({ hasError }: Props): number => (hasError ? -60 : -40)}px;
  border-radius: 20px;
  padding: 5px;
  background-color: ${({ theme }: ThemeContextType): string => theme.monochromatic.white};
`;

export const HelperText = styled.View`
  margin-bottom: 2px;
`;
