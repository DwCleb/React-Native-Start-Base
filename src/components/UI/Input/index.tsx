import * as React from 'react';
import { TextInputMask } from 'react-native-masked-text';
import { Icon } from '~/components';
import { useTheme } from '~/theme';

import Typography, { TypographyType } from '~/components/UI/Typography';
import { InnerInput, ShowPass, InputMode, InputBase, HelperText } from './styles';
// -> import enums and props
import { InputProps, PassIcon } from './styles';

export default function Input({
  label,
  value,
  name,
  onChange,
  pass,
  paddingHorizontal = 20,
  mode = InputMode.OUTLINED,
  height,
  distance,
  helperText,
  options,
  mask,
  onBlur,
  onFocus,
}: InputProps): JSX.Element {
  const { theme } = useTheme();
  const [showPassword, setShowPassword] = React.useState(pass);

  function returnMaskedInput(): JSX.Element {
    return (
      <InputBase
        render={(props: any): JSX.Element => <TextInputMask {...props} type={mask} options={options} />}
        label={label}
        height={height}
        mode={mode}
        value={value}
        hasError={helperText}
        secureTextEntry={showPassword}
        onChangeText={(textValue: string): void => onChange({ value: textValue, name })}
        onBlur={onBlur}
        onFocus={onFocus}
      />
    );
  }

  function returnSimpleInput(): JSX.Element {
    return (
      <InputBase
        label={label}
        height={height}
        mode={mode}
        value={value}
        hasError={helperText}
        secureTextEntry={showPassword}
        onChangeText={(textValue: string): void => onChange({ value: textValue, name })}
        onBlur={onBlur}
        onFocus={onFocus}
      />
    );
  }

  return (
    <InnerInput paddingHorizontal={paddingHorizontal} distance={distance} hasError={helperText}>
      {/* render the correct input by user need */}
      {mask ? returnMaskedInput() : returnSimpleInput()}

      {/* component helper text to show warning and error messages */}
      {helperText && (
        <HelperText>
          <Typography type={TypographyType.L1} color={theme.alerts.red} align="left">
            {helperText}
          </Typography>
        </HelperText>
      )}

      {/* component showpass to input entry pass  */}
      {pass && (
        <ShowPass onPress={(): void => setShowPassword(!showPassword)} hasError={helperText}>
          <Icon name={showPassword ? PassIcon.OFF : PassIcon.ON} color={theme.primary.blue} />
        </ShowPass>
      )}
    </InnerInput>
  );
}
