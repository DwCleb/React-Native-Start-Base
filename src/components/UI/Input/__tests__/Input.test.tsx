import React from 'react';
import { View, Text } from 'react-native';

/* Testing Library */
import { create, act, ReactTestRenderer } from 'react-test-renderer';

/* Component to test */
import Input, { InputProps } from '../';
import { TextInput, LabelText, InputLine, InputContainer } from '../styles';
import { InputVariants, InputStatus } from '../../Input/InputType';

/* Theme */
import { ThemeProvider } from 'styled-components/native';
import darkTheme from '~/theme/ThemeContext/dark';

/* Assets */
import { CartIcon, CheckIcon } from '~/assets';

let wrapper: ReactTestRenderer;
let root: ReactTestRenderer['root'];

const onChangeEvent = jest.fn();
const onTouchEventMock = jest.fn();

let props: InputProps = {
  name: 'inputTest',
  onChange: onChangeEvent,
  value: '',
};

const { theme } = darkTheme;

const componentToTest = (componentProps: InputProps): JSX.Element => (
  <ThemeProvider theme={theme}>
    <Input {...componentProps} />
  </ThemeProvider>
);

describe('<Input />', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = create(componentToTest(props));
    root = wrapper.root;
  });

  it('Should render correctly', () => {
    props = { ...props };
    expect(wrapper).toMatchSnapshot();
  });

  it('Should the correct structure', () => {
    expect(root.findAllByType(TextInput)).toHaveLength(1);

    expect(root.findAllByType(InputContainer)).toHaveLength(1);

    //default variant should be outlined
    expect(root.findAllByType(InputContainer)[0].props.variant).toBe(InputVariants.OUTLINED);
  });

  it('Should call onChangeEvent if text is changed', () => {
    const newProps = {
      ...props,
      value: '12',
    };
    wrapper = create(componentToTest(newProps));
    root = wrapper.root;

    const textInputElement = root.findAllByType(TextInput)[0];

    act(() => {
      textInputElement.props.onChangeText('12');
    });

    expect(onChangeEvent).toHaveBeenCalled();
  });

  it('Should change inputState to default with onFocus and value on input', () => {
    const newProps = {
      ...props,
      value: '12',
    };
    wrapper = create(componentToTest(newProps));
    root = wrapper.root;

    const textInputElement = root.findAllByType(TextInput)[0];

    act(() => {
      textInputElement.props.onFocus();
    });

    expect(root.findAllByType(InputContainer)[0].props.inputStatus).toBe(InputStatus.Initial);
  });

  it('If prop inputStatus is Complete, the rendered components should also be Complete', () => {
    const newProps = {
      ...props,
      value: '123',
      inputStatus: InputStatus.Complete,
    };
    wrapper = create(componentToTest(newProps));
    root = wrapper.root;

    expect(root.findAllByType(InputContainer)[0].props.inputStatus).toBe(InputStatus.Complete);
  });

  it('Should render a label if prop is received and has a value', () => {
    const newProps = {
      ...props,
      label: 'testInput',
      value: '123',
    };

    wrapper = create(componentToTest(newProps));
    root = wrapper.root;

    expect(root.findAllByType(LabelText).length).toBe(1);
  });

  it('Should render as regular variant, if prop received', () => {
    const newProps = {
      ...props,
      variant: InputVariants.REGULAR,
    };

    wrapper = create(componentToTest(newProps));
    root = wrapper.root;

    expect(root.findAllByType(InputContainer)[0].props.variant).toBe(InputVariants.REGULAR);
    expect(root.findAllByType(InputLine).length).toBe(1);
  });

  it('Should render as invalid, if isInvalid prop is true', () => {
    const newProps = {
      ...props,
      value: '123',
      inputStatus: InputStatus.Invalid,
    };

    wrapper = create(componentToTest(newProps));
    root = wrapper.root;

    expect(root.findAllByType(InputContainer)[0].props.inputStatus).toBe(InputStatus.Invalid);
  });

  it('Should render as valid, if isValid prop is true', () => {
    const newProps = {
      ...props,
      value: '123',
      leftIcon: { source: CartIcon, validSource: CheckIcon },
      inputStatus: InputStatus.Valid,
    };
    wrapper = create(componentToTest(newProps));
    root = wrapper.root;

    expect(root.findAllByType(InputContainer)[0].props.inputStatus).toBe(InputStatus.Valid);
  });

  it('OnBlur event should trigger onTouchEvent prop, if prop is received', () => {
    const newProps = {
      ...props,
      onTouchEvent: onTouchEventMock,
    };

    wrapper = create(componentToTest(newProps));
    root = wrapper.root;

    const textInputElement = root.findAllByType(TextInput)[0];

    act(() => {
      textInputElement.props.onBlur();
    });

    expect(onTouchEventMock).toHaveBeenCalled();
  });

  it('OnFocus event should trigger onTouchEvent prop, if prop is received', () => {
    const newProps = {
      ...props,
      onTouchEvent: onTouchEventMock,
    };

    wrapper = create(componentToTest(newProps));
    root = wrapper.root;

    const textInputElement = root.findAllByType(TextInput)[0];

    act(() => {
      textInputElement.props.onFocus();
    });

    expect(onTouchEventMock).toHaveBeenCalled();
  });

  it('it should not trigger onChange event, if no onChange Prop is Provided', () => {
    const newProps = {
      ...props,
      onChange: undefined,
    };

    wrapper = create(componentToTest(newProps));
    root = wrapper.root;

    const textInputElement = root.findAllByType(TextInput)[0];

    act(() => {
      textInputElement.props.onChangeText('1');
    });

    expect(onChangeEvent).toHaveBeenCalledTimes(0);
  });

  it('it should render right and left elements if provided', () => {
    const newProps = {
      ...props,
      leftElement: (
        <View>
          <Text>LEFT</Text>
        </View>
      ),
      rightElement: (
        <View>
          <Text>RIGHT</Text>
        </View>
      ),
    };

    wrapper = create(componentToTest(newProps));
    root = wrapper.root;

    expect(root.findAllByType(InputContainer)[0].props.children[0]).toEqual(newProps.leftElement);
    expect(root.findAllByType(InputContainer)[0].props.children[3]).toEqual(newProps.rightElement);
  });
});
