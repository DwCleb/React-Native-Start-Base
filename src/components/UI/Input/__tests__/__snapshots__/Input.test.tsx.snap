// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> Should render correctly 1`] = `
<View
  inputStatus={0}
  style={
    Array [
      Object {
        "alignSelf": "center",
        "borderColor": "#FFFFFF4D",
        "borderRadius": 16,
        "borderWidth": 4,
        "flexDirection": "row",
        "height": 110,
        "overflow": "hidden",
        "position": "relative",
        "width": "100%",
      },
    ]
  }
  value=""
  variant={1}
>
  <View
    inputStatus={0}
    leftBorder={false}
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
          "paddingBottom": 0,
          "paddingLeft": 40,
          "paddingRight": 40,
          "paddingTop": 0,
        },
      ]
    }
    value=""
    variant={1}
  >
    <TextInput
      accessibilityLabel="inputTest"
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      inputStatus={0}
      keyboardAppearance="dark"
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#FFFFFF99"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 28,
            "height": "100%",
            "lineHeight": 42,
            "paddingBottom": 0,
            "paddingLeft": 0,
            "paddingRight": 0,
            "paddingTop": 0,
            "textAlign": "left",
            "width": "100%",
          },
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;
